### PR TITLE
SPIR-V Runner: Integrate argument parsing library

### PR DESCRIPTION
**Pull request by kballeda**
This PR integrates LLVM based argument parsing library into SPIR-V Runner and closes https://github.com/intel/intel-xpu-backend-for-triton/issues/2568

```
/intel-xpu-backend-for-triton/utils/SPIRVRunner$ ./build/SPIRVRunner -o tensor_2
Running on device: Intel(R) Data Center GPU Max 1100
Read 3772 byte kernel.
Loaded kernel with 0 registers and 0 register spills.
Tensor output: [98432], Float (393728 bytes)
Output Tensor Path: intel-xpu-backend-for-triton/utils/SPIRVRunner/cpp_outs.pt

intel-xpu-backend-for-triton/utils/SPIRVRunner$ ./build/SPIRVRunner -o tensor_2 -p
Running on device: Intel(R) Data Center GPU Max 1100
Read 3772 byte kernel.
Loaded kernel with 0 registers and 0 register spills.
Tensor output: [98432], Float (393728 bytes)
Kernel execution time: 0.0096 ms
Output Tensor Path: intel-xpu-backend-for-triton/utils/SPIRVRunner/cpp_outs.pt

intel-xpu-backend-for-triton/utils/SPIRVRunner$ ./build/SPIRVRunner --help
USAGE: SPIRVRunner [options]

OPTIONS:

Color Options:

  --color       - Use colors in output (default=autodetect)

General options:

  -o <filename> - Specify Output Tensor Name
  -p            - Enable Profiling


```

### CODE REVIEW

File: utils/SPIRVRunner/CMakeLists.txt [hunk 1]
```diff
@@ -38,10 +44,10 @@ set(SYCL_FUNCTIONS_INCLUDE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/../../third_party/in
...
 add_executable(${TARGET_NAME} ${TARGET_NAME}.cpp)
 target_include_directories(${TARGET_NAME} PRIVATE
-    "${ONEAPI_ROOT}/compiler/latest/include" ${SYCL_FUNCTIONS_INCLUDE_DIR} ${JSON_INCLUDE_DIR})
+	"${ONEAPI_ROOT}/compiler/latest/include" ${SYCL_FUNCTIONS_INCLUDE_DIR} ${JSON_INCLUDE_DIR} ${LLVM_INCLUDE_DIRS})
 set_target_properties(${TARGET_NAME} PROPERTIES COMPILE_FLAGS "${COMPILE_FLAGS}")
 set_target_properties(${TARGET_NAME} PROPERTIES LINK_FLAGS "${LINK_FLAGS}")
 add_dependencies(${TARGET_NAME} json)
 
-target_link_libraries(${TARGET_NAME} "${TORCH_LIBRARIES}")
+target_link_libraries(${TARGET_NAME} PRIVATE "${TORCH_LIBRARIES}" "${LLVM_LIBS}")
```
[the remark this answers is not in the document]
Comment (kballeda) [reply utils/SPIRVRunner/CMakeLists.txt:52]: Thank you for the review. As discussed offline, https://github.com/intel/intel-xpu-backend-for-triton/issues/2568 necessitates the use of the argument parsing library that is already part of the current codebase. We have determined that LLVM is suitable for integration and is also available as a static library (LLVMSupport)

**Review by anmyachev — APPROVED** [review]
LGTM!

File: scripts/compile-triton.sh [hunk 2]
```diff
@@ -129,6 +129,7 @@ build_llvm() {
   cd $LLVM_PROJ_BUILD
   cmake -G Ninja ../llvm \
     -DLLVM_ENABLE_DUMP=1 \
+    -DLLVM_ENABLE_RTTI=ON \
```
[the remark this answers is not in the document]
Comment (kballeda) [reply scripts/compile-triton.sh:132]: I get following error:

<details>
  <summary>Click to expand</summary>
[100%] Linking CXX executable SPIRVRunner
/usr/bin/ld: /tmp/icpx-fb3330f94f/SPIRVRunner-445ef4.o:(.rodata._ZTIN4llvm2cl15OptionValueCopyINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEE[_ZTIN4llvm2cl15OptionValueCopyINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEE]+0x10): undefined reference to `typeinfo for llvm::cl::GenericOptionValue'
/usr/bin/ld: /tmp/icpx-fb3330f94f/SPIRVRunner-445ef4.o:(.rodata._ZTIN4llvm2cl15OptionValueCopyIbEE[_ZTIN4llvm2cl15OptionValueCopyIbEE]+0x10): undefined reference to `typeinfo for llvm::cl::GenericOptionValue'
icpx: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [CMakeFiles/SPIRVRunner.dir/build.make:110: SPIRVRunner] Error 1


</details>

[the remark this answers is not in the document]
Comment (alexbaden) [reply scripts/compile-triton.sh:132]: Can you try building the SPIRV Runner w/out RTTI?

[the remark this answers is not in the document]
Comment (kballeda) [reply scripts/compile-triton.sh:132]: I have used LLVM (no RTTI default/similar to log above) + disabled RTTI in SPIRVRunner cmake file with following,

`target_compile_option(${TARGET_NAME} PRIVATE -fno-rtti)`

Error Log:

<details>
<summary>Click to show log</summary>

/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:208:61: error: use of typeid requires -frtti
  208 |         CodeLoc, (void *)impl.get(), nullptr, (const void *)typeid(T).name(),
      |                                                             ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:221:61: error: use of typeid requires -frtti
  221 |         CodeLoc, (void *)impl.get(), nullptr, (const void *)typeid(T).name(),
      |                                                             ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:234:62: error: use of typeid requires -frtti
  234 |         CodeLoc, (void *)impl.get(), hostData, (const void *)typeid(T).name(),
      |                                                              ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:247:62: error: use of typeid requires -frtti
  247 |         CodeLoc, (void *)impl.get(), hostData, (const void *)typeid(T).name(),
      |                                                              ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:262:62: error: use of typeid requires -frtti
  262 |         CodeLoc, (void *)impl.get(), hostData, (const void *)typeid(T).name(),
      |                                                              ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:277:62: error: use of typeid requires -frtti
  277 |         CodeLoc, (void *)impl.get(), hostData, (const void *)typeid(T).name(),
      |                                                              ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:293:23: error: use of typeid requires -frtti
  293 |         (const void *)typeid(T).name(), dimensions, sizeof(T),
      |                       ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:309:23: error: use of typeid requires -frtti
  309 |         (const void *)typeid(T).name(), dimensions, sizeof(T),
      |                       ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:325:23: error: use of typeid requires -frtti
  325 |         (const void *)typeid(T).name(), dimensions, sizeof(T),
      |                       ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:341:23: error: use of typeid requires -frtti
  341 |         (const void *)typeid(T).name(), dimensions, sizeof(T),
      |                       ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:373:57: error: use of typeid requires -frtti
  373 |                                           (const void *)typeid(T).name(),
      |                                                         ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:405:57: error: use of typeid requires -frtti
  405 |                                           (const void *)typeid(T).name(),
      |                                                         ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:425:23: error: use of typeid requires -frtti
  425 |         (const void *)typeid(T).name(), dimensions, sizeof(T), r);
      |                       ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:443:64: error: use of typeid requires -frtti
  443 |         CodeLoc, (void *)impl.get(), impl.get(), (const void *)typeid(T).name(),
      |                                                                ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:462:64: error: use of typeid requires -frtti
  462 |         CodeLoc, (void *)impl.get(), impl.get(), (const void *)typeid(T).name(),
      |                                                                ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:471:64: error: use of typeid requires -frtti
  471 |         CodeLoc, (void *)impl.get(), impl.get(), (const void *)typeid(T).name(),
      |                                                                ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:761:64: error: use of typeid requires -frtti
  761 |         CodeLoc, (void *)impl.get(), &MemObject, (const void *)typeid(T).name(),
      |                                                                ^
/opt/intel/oneapi/compiler/2025.0/include/sycl/buffer.hpp:781:64: error: use of typeid requires -frtti
  781 |         CodeLoc, (void *)impl.get(), Impl.get(), (const void *)typeid(T).name(),
      |                                                                ^
In file included from /home/jovyan/Kali/1119_args_new/intel-xpu-backend-for-triton/utils/SPIRVRunner/SPIRVRunner.cpp:2:
In file included from /opt/intel/oneapi/compiler/2025.0/include/sycl/sycl.hpp:27:
/opt/intel/oneapi/compiler/2025.0/include/sycl/accessor_image.hpp:980:35: error: use of typeid requires -frtti
  980 |         AccessMode, (const void *)typeid(DataT).name(), sizeof(DataT), CodeLoc);
      |                                   ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]

</details>

[the remark this answers is not in the document]
Comment (alexbaden) [reply scripts/compile-triton.sh:132]: Ah, ok. I think it's worth trying an experiment to see if -fno-rtti will work here. Split the command line options parsing out to a separate file (without sycl) and see if that compiles, links, and runs with `-fno-rtti`. If it does, we might be able to build the command line parsing as a separate object file and then link that object file with our main `SPIRVRunner` object file to get around this problem.

**Comment by etiotto** [comment]
@alexbaden are you ok with these changes ?

**Review by alexbaden — CHANGES_REQUESTED** [review]
No - see comment: https://github.com/intel/intel-xpu-backend-for-triton/pull/2759#discussion_r1852550872

File: scripts/compile-triton.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (kballeda) [reply scripts/compile-triton.sh]: I was out for some time. yes, this change was experimented, and it works. Updated the PR, please review the same. Thanks

**Review by alexbaden — APPROVED** [review]
Looks good, happy to see the `-fno-rtti` trick worked!

File: utils/SPIRVRunner/llvm_parser.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#ifndef LLVM_PARSER_H
+#define LLVM_PARSER_H
+
+#include "llvm/Support/CommandLine.h"
+
+class command_line_parser {
+public:
+  struct options {
+    std::string output_tensor;
```
[the remark this answers is not in the document]
Comment (kballeda) [reply utils/SPIRVRunner/llvm_parser.h:9]: Sure, will check this!

[the remark this answers is not in the document]
Comment (alexbaden) [reply utils/SPIRVRunner/llvm_parser.h:9]: Looks good, thanks!

**Pull request merged**